qtfred/src/CheckedDataListModel.h

```diff
@@ -141,21 +141,84 @@ class CheckedDataListModel : public QAbstractListModel
 			return defaultFlags;
 	}
 
-	inline const SCP_unordered_set<T*>& getCheckedData() {
-		return checkeds;
+private:
+	template<typename V>
+	inline void getCheckedDataInternal(SCP_unordered_set<V*> &set) {
+		for (auto& item : items)
+			if (item._checked)
+				set.insert(&item.internalData());
 	}
 
-	inline const SCP_unordered_set<const T*>& getCheckedDataConst() const {
-		return checkedsConst;
+public:
+	inline SCP_unordered_set<T*> getCheckedData() {
+		SCP_unordered_set<T*> ret{items.size() / 2};
+		getCheckedDataInternal(ret);
+		return ret;
 	}
 
+	inline SCP_unordered_set<const T*> getCheckedData() const {
+		SCP_unordered_set<const T*> ret{items.size() / 2};
+		const_cast<CheckedDataListModel<T>*>(this)->getCheckedDataInternal(ret);
+		return ret;
+	}
+
+	struct Iterator
+	{
+		using iterator_delegate = typename SCP_vector<RowData<T>>::iterator;
+
+		using iterator_category = std::forward_iterator_tag;
+		using difference_type   = typename SCP_vector<RowData<T>>::iterator::difference_type;
+		using value_type        = std::pair<T, bool>;
+		using pointer           = std::pair<const T*, bool>;
+		using reference         = std::pair<T&, bool>;
+
+		Iterator(iterator_delegate it) : _it(it) {}
+
+		reference operator*() const { return reference(_it->internalData(), _it->_checked); }
+		pointer operator->() { return pointer(&_it->internalData(), _it->_checked); }
+		Iterator& operator++() { _it++; return *this; }
+		Iterator operator++(int) { Iterator tmp = *this; ++(*this); return tmp; }
+		friend bool operator== (const Iterator &a, const Iterator &b) { return a._it == b._it;}
+		friend bool operator!= (const Iterator &a, const Iterator &b) { return a._it != b._it;}
```

Comment:
I'm no longer quite sure why they needed to be friends instead of members. They're part of what an input iterator, and as such a forward iterator, needs to provide.

Reply:
I'm not doing anything fancy with these iterators (except iterating), so I certainly don't mind changing them.